akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala

```diff
@@ -0,0 +1,399 @@
+   */
+  def spawnAdapter[U](f: U ⇒ T, name: String): ActorRef[U]
+
+  /**
+   * Create an anonymous child actor that will wrap messages such that other Actor’s
+   * protocols can be ingested by this Actor. You are strongly advised to cache
+   * these ActorRefs or to stop them when no longer needed.
+   */
+  def spawnAdapter[U](f: U ⇒ T): ActorRef[U] = spawnAdapter(f, "")
+
+}
+
+object Actor {
+  import Behavior._
+
+  // FIXME check that all behaviors can cope with not getting PreStart as first message
+
+  final implicit class BehaviorDecorators[T](val behavior: Behavior[T]) extends AnyVal {
+    /**
+     * Widen the wrapped Behavior by placing a funnel in front of it: the supplied
+     * PartialFunction decides which message to pull in (those that it is defined
+     * at) and may transform the incoming message to place them into the wrapped
+     * Behavior’s type hierarchy. Signals are not transformed.
+     *
+     * see also [[Actor.Widened]]
+     */
+    def widen[U](matcher: PartialFunction[U, T]): Behavior[U] = Widened(behavior, matcher)
+  }
+
+  private val _nullFun = (_: Any) ⇒ null
+  private def nullFun[T] = _nullFun.asInstanceOf[Any ⇒ T]
+  private implicit class ContextAs[T](val ctx: AC[T]) extends AnyVal {
+    def as[U] = ctx.asInstanceOf[AC[U]]
+  }
+
+  /**
+   * Widen the wrapped Behavior by placing a funnel in front of it: the supplied
+   * PartialFunction decides which message to pull in (those that it is defined
+   * at) and may transform the incoming message to place them into the wrapped
+   * Behavior’s type hierarchy. Signals are not transformed.
+   *
+   * Example:
+   * {{{
+   * Stateless[String]((ctx, msg) => println(msg)).widen[Number] {
+   *   case b: BigDecimal => s"BigDecimal($b)"
+   *   case i: BigInteger => s"BigInteger($i)"
+   *   // drop all other kinds of Number
+   * }
+   * }}}
+   */
+  final case class Widened[T, U](behavior: Behavior[T], matcher: PartialFunction[U, T]) extends Behavior[U] {
+    private def postProcess(behv: Behavior[T]): Behavior[U] =
+      if (isUnhandled(behv)) Unhandled
+      else if (isAlive(behv)) {
+        val next = canonicalize(behv, behavior)
+        if (next eq behavior) Same else Widened(next, matcher)
+      } else Stopped
+
+    override def management(ctx: AC[U], msg: Signal): Behavior[U] =
+      postProcess(behavior.management(ctx.as[T], msg))
+
+    override def message(ctx: AC[U], msg: U): Behavior[U] =
+      matcher.applyOrElse(msg, nullFun) match {
+        case null        ⇒ Unhandled
+        case transformed ⇒ postProcess(behavior.message(ctx.as[T], transformed))
+      }
+
+    override def toString: String = s"${behavior.toString}.widen(${LineNumbers(matcher)})"
+  }
+
+  /**
+   * Wrap a behavior factory so that it runs upon PreStart, i.e. behavior creation
+   * is deferred to the child actor instead of running within the parent.
+   */
+  final case class Deferred[T](factory: ActorContext[T] ⇒ Behavior[T]) extends Behavior[T] {
```

Comment:
well, it matches semantically, it is all just data, and it allows pattern matching and decomposition